File: Day41/DelNodeInBST.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode(int x) { val = x; }
}

public class DelNodeInBST {
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) return null;
        if (key < root.val) {
            root.left = deleteNode(root.left, key);
        } else if (key > root.val) {
            root.right = deleteNode(root.right, key);
        } else {
            if (root.left == null) return root.right;
            if (root.right == null) return root.left;
            TreeNode temp = root.right;
            while (temp.left != null) {
                temp = temp.left;
            }
            root.val = temp.val;
            root.right = deleteNode(root.right, temp.val);
        }
        return root;
    }

    public static TreeNode buildTree(String[] vals) {
        if (vals.length == 0 || vals[0].equals("null") || vals[0].equals("")) return null;
        TreeNode root = new TreeNode(Integer.parseInt(vals[0].trim()));
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int i = 1;
        while (!queue.isEmpty() && i < vals.length) {
            TreeNode node = queue.poll();
            if (i < vals.length && !vals[i].trim().equals("null") && !vals[i].trim().equals("")) {
                node.left = new TreeNode(Integer.parseInt(vals[i].trim()));
                queue.offer(node.left);
            }
            i++;
            if (i < vals.length && !vals[i].trim().equals("null") && !vals[i].trim().equals("")) {
                node.right = new TreeNode(Integer.parseInt(vals[i].trim()));
                queue.offer(node.right);
            }
            i++;
        }
        return root;
    }

    public static void printBST(TreeNode root) {
        if (root == null) return;
        printBST(root.left);
        System.out.print(root.val + " ");
        printBST(root.right);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String input = sc.nextLine().replaceAll("[\\[\\]]", "");
        String[] vals = input.split(",");
        int key = sc.nextInt();
        TreeNode root = buildTree(vals);
        DelNodeInBST sol = new DelNodeInBST();
        TreeNode result = sol.deleteNode(root, key);
        printBST(result);
        System.out.println();
    }
}

File: Day41/DelNodeInBST.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def deleteNode(self, root, key):
        if not root:
            return None
        if key < root.val:
            root.left = self.deleteNode(root.left, key)
        elif key > root.val:
            root.right = self.deleteNode(root.right, key)
        else:
            if not root.left:
                return root.right
            if not root.right:
                return root.left
            temp = root.right
            while temp.left:
                temp = temp.left
            root.val = temp.val
            root.right = self.deleteNode(root.right, temp.val)
        return root

    @staticmethod
    def build_tree(nodes):
        vals = nodes.replace('[','').replace(']','').split(',')
        if not vals or vals[0].strip() == 'null' or vals[0].strip() == '':
            return None
        root = TreeNode(int(vals[0]))
        queue = [root]
        i = 1
        while queue and i < len(vals):
            node = queue.pop(0)
            if i < len(vals) and vals[i].strip() != 'null' and vals[i].strip() != '':
                node.left = TreeNode(int(vals[i]))
                queue.append(node.left)
            i += 1
            if i < len(vals) and vals[i].strip() != 'null' and vals[i].strip() != '':
                node.right = TreeNode(int(vals[i]))
                queue.append(node.right)
            i += 1
        return root

    @staticmethod
    def printBST(root):
        if root:
            Solution.printBST(root.left)
            print(root.val, end=" ")
            Solution.printBST(root.right)

    @staticmethod
    def main():
        tree = input().strip()
        key = int(input().strip())
        root = Solution.build_tree(tree)
        sol = Solution()
        result = sol.deleteNode(root, key)
        Solution.printBST(result)
        print()

if __name__ == "__main__":
    Solution.main()
